sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -247,58 +259,137 @@ object CTESubstitution extends Rule[LogicalPlan] {
         // NOTE: we must call `traverseAndSubstituteCTE` before `substituteCTE`, as the relations
         // in the inner CTE have higher priority over the relations in the outer CTE when resolving
         // inner CTE relations. For example:
-        // WITH t1 AS (SELECT 1)
-        // t2 AS (
-        //   WITH t1 AS (SELECT 2)
-        //   WITH t3 AS (SELECT * FROM t1)
-        // )
+        // WITH
+        //   t1 AS (SELECT 1),
+        //   t2 AS (
+        //     WITH
+        //       t1 AS (SELECT 2),
+        //       t3 AS (SELECT * FROM t1)
+        //     SELECT * FROM t1
+        //   )
+        // SELECT * FROM t2
         // t3 should resolve the t1 to `SELECT 2` instead of `SELECT 1`.
-        traverseAndSubstituteCTE(relation, forceInline, resolvedCTERelations, cteDefs)._1
+        //
+        // When recursion allowed:
+        // - don't add current definition to outer definitions of `traverseAndSubstituteCTE()` to
+        //   prevent recursion inside inner CTEs.
+        //   E.g. the following query will not resolve `t1` within `t2`:
+        //   WITH RECURSIVE
+        //     t1 AS (
+        //       SELECT 1 AS level
+        //       UNION (
+        //         WITH t2 AS (SELECT level + 1 FROM t1 WHERE level < 10)
+        //         SELECT * FROM t2
+        //       )
+        //     )
+        //   SELECT * FROM t1
+        // - remove definitions that conflict with current relation `name` from outer definitions of
+        //   `traverseAndSubstituteCTE()` to prevent weird resolutions.
+        //   E.g. we don't want to resolve `t1` within `t3` to `SELECT 1`:
+        //   WITH
+        //     t1 AS (SELECT 1),
+        //     t2 AS (
+        //       WITH RECURSIVE
+        //         t1 AS (
+        //           SELECT 1 AS level
+        //           UNION (
+        //             WITH t3 AS (SELECT level + 1 FROM t1 WHERE level < 10)
```

Comment:
how is it different from the previous note? It seems to be the same thing: recursive cte relations can be referenced by nested inner CTEs.

Reply:
oh I see, it also removes outer CTE relations that have the same name as the recursive CTE relations

Reply:
New comments should explain this in a better way